### ISSUE TITLE
Migrate AOI creation process to BAGIS-Pro

### ISSUE THREAD

**Issue by lbross** [PSU-CSAR/bagis-pro#53]
Start with create AOI from shapefile.
Clips all the layers and arrange them into an AOI structure. Creates the flow/accumulation layers. New step for tool to create a pourpoint. Intersection with stream network should be the pourpoint. Find location inside shapefile that has the highest allocation value. 
Interactive tool user needs to digitize a pourpoint. Make decision on which gauge station to use.

**Comment by lbross** [comment]
@jdduh: Can you provide a sample shapefile for me to work with, hopefully not too big?

**Comment by jdduh** [comment]
Here is a shapefile for the Animas AOI.

[animas_aoi.zip](https://github.com/user-attachments/files/16988917/animas_aoi.zip)

**Comment by lbross** [comment]
Thank you. I'm pasting a screenshot of the UI for this function. I plan to replicate the screen as much as possible. Let me know if you would like to make any changes as we proceed with the transition.
![image](https://github.com/user-attachments/assets/314ef88a-efb3-4a5e-88c9-32aebdadfbeb)

**Comment by jdduh** [comment]
We need a new BAGIS tool 'group' for 'AOI Creation'. All tools, including the 'create AOI from a shapefile' tool, in the Basin Analyst pulldown menu can be moved to the tool group. 
![image](https://github.com/user-attachments/assets/87b78ebc-6f9e-4b1a-9b12-c2fadbd2b3db)
![image](https://github.com/user-attachments/assets/0e3b6d42-8a1d-46f2-963e-009907610dbd)

**Comment by lbross** [comment]
This is what I'm working with right now. I created the Basin Analyst tool group a couple of years ago when we started working with ArcGIS Pro. I need a way to trigger the tool. I'm going to focus on the tool for now, but will update the menu before release.
![image](https://github.com/user-attachments/assets/b23aa589-a2d7-447f-bdef-bf8ecfd92a38)

**Comment by lbross** [comment]
This is what the UI looks like converted to BAGIS-Pro. No code behind it yet.
![image](https://github.com/user-attachments/assets/f857a31e-1823-4e87-b1b8-b5e7446f4991)

**Comment by lbross** [comment]
Just noticed that the height and width of the filter size is always disabled in BAGIS v3. ie: if you choose to filter, it will always be 3x7. Is this intentional?

**Comment by jdduh** [comment]
The Basin Analysis tool group looks good. In this case, we can keep the "About" menu item and "hide" the other menu items that I crossed out. We probably will convert some of the AOI utilities functions to BAGIS Pro later. 

The filter size should be customizable by the users. 3 x 7 is the recommended size. The default for DEM smoothing is disable. Here is the description on the setting.
![image](https://github.com/user-attachments/assets/3463e4f3-2e5d-4913-9170-b206ec9a0b6b)

**Comment by lbross** [comment]
Per our conversation on Tuesday, I'd already made the changes to the menu that you requested earlier. I can make additional changes as needed. This is what I currently have:
![image](https://github.com/user-attachments/assets/098c4eef-6979-4178-85b5-a513e7c8de8c)
I've migrated the 'Why Smooth DEM' MessageBox to BAGIS-Pro and made the filter size customizable.
I'm running into some logic that is dependent on the 'Generate AOI Only' checkbox on the BAGIS V3 settings screen. Is this something we still need to support? It seems like an AOI isn't very usable with the BAGIS-Pro reporting if it doesn't have all of the layers.

**Comment by lbross** [comment]
We need to think about where this tool should get the uri's for the clipping. We read the bagis-pro data sources and their uri's from the GitHub server. There is no way to override these, but we do keep a record of them locally for each aoi to include in the pdf report. We also have the default values for the basinanalyst.def online via GitHub and use these a few places in bagis-pro. Then there is the local basinanalyst.def file. BAGIS V3 gets its uris from there. This allows the analyst to override the default settings by using the 'Options' menu. They can also synchronize the options with the default settings at any time.

**Comment by lbross** [comment]
There is a function to add the aoiname, basin, and awdb_id fields to the aoi_v feature class? Are these still the correct fields? Should we be adding station_triplet?

**Comment by jdduh** [comment]
When creating an AOI from a shapefile, the 'station' information is not required. When creating an AOI based on station information, users must create a BASIN first to subset the DEM so that the flowdir and flowacc can be calculated efficiently. Once the BASIN exists, users can create an AOI (within the BASIN) based on an existing station or a location (non-station) that they picked. The stationName and stationTriplet data are available in one of these situations. For other situations, users still need to provide a name for the AOI. This name is not a stationName. When creating an AOI from a BASIN, the BASIN name is propagated to basin attribute of the AOI.

I suggest that, for the create an AOI from a shapefile tool, we keep the original program logics. The AOIs created with this tool need additional methods to find their pourpoints. Once we have the pourpoints, we can then integrate the 'Update forecast station data' tool to see if any of the pourpoints falls close to existing forecast stations.

For AOIs created from a BASIN, we can use the station data or the default station data, when the station doesn't exist, to set the stationName and stationTriplet values and ignore the aoiname and awdb_id fields. Let me know if you have any thoughts.

**Comment by lbross** [comment]
The original program logic for this tool is to add the aoiname, basin, and awdb_id fields to the aoi_v feature class immediately after the feature class is created. As you indicated, there is only enough information available to populate the aoiname field. The other two remain null. An aoi_v feature class should have the same attribute fields regardless of how it was created. They may not all be populated.
My only suggestion is to stop creating awdb_id and start creating station_triplet since we no longer use awdb_id. We might also want to start adding station_name at this time.

**Comment by jdduh** [comment]
Let's ditch aoiname and awdb_id field and keep basin. We will set Nodata (or whatever we decided on the default value) on stationTriplet and the AOI name users entered as the stationName for the create an AOI from a shapefile tool. We can just leave the basin name blank.

**Comment by lbross** [comment]
To summarize: For this tool we will add 3 fields to aoi_v when it is created: stationName, stationTriplet, and basin. For this tool, StationName will be set to the value the user enters in the 'Output AOI Name" field. This could be overwritten by the forecast station tool or an integrated version of that logic. The remaining 2 fields will be left null.

**Comment by lbross** [comment]
Add new fields to batch_settings_tool.json for dem uri, the dem units, and the gauge station layers. Rename BatchSettings to BagisSettings. Give these 3 new data points a unique prefix so that they can be easily found. Make sure we are recording the buffer distance and units for the dem (AoiBufferUnits, AoiBufferDistance). Remove references to PRISM buffer and units from UI.

**Comment by lbross** [comment]
I am in the process of renaming BatchSettings to BagisSettings. This will break BAGIS-Pro. Let me know if you need to use BAGIS-Pro and I will post a new version

**Comment by jdduh** [comment]
Please post a new version when you get a chance. I need to generate some new basin reports. Thanks!

**Comment by lbross** [comment]
There is a new version posted to the basins ftp server. It is lightly tested, so let me know if you run into any issues. If I can't fix them quickly, we can change the server back until you generate the new reports. The names of the fields with the years in the fire statistics tool are probably still not right. I am waiting for confirmation from you on issue #55

**Comment by jdduh** [comment]
The batch tool reporting function works normally without a glitch. The pdf files were created correctly. I haven't tried the summary tool yet.

**Comment by jdduh** [comment]
Except some scrambled codes on the cover page, probably related to #54.

![image](https://github.com/user-attachments/assets/41fe4e71-39c5-4124-b416-d9891234fd39)

and this (the scrambled codes show up in different places in different documents)

![image](https://github.com/user-attachments/assets/53313e1e-a6de-4451-baa9-fb0405025d40)

The pdfs can be found at: ftp://webservices.geog.pdx.edu/public/reports/Oct2024_NewForecastStations/

**Comment by lbross** [comment]
Yes. This is related to #54 and I don't know how soon or if it can be fixed. The other free PDF generators I've explored have issues of their own. Maybe we can come up with an alternate to the dash (-)? This is Chrome trying to localize when we haven't built a profile.

**Comment by jdduh** [comment]
We can replace the long horizontal bars with dashes. I'm not sure about the carriage return characters that appear between the paragraphs.

**Comment by lbross** [comment]
So this is interesting. It looks fine on my laptop. What computer are you running the tool on? Do you have some of the international features enabled for Chrome? What version of Chrome is installed?

**Comment by jdduh** [comment]
Windows 11 22H2 with Chinese language pack installed, but all the default language settings are English (US); Chrome version 130.0.6723.70 (just updated, but I check for Chrome updates fairly frequently). Chrome doesn't have new extensions installed recently. All the extensions were disabled.

**Comment by jdduh** [comment]
This zip file contains the source html and kml files of the rendered pdf pages and the two report pdfs (with the first 4 pages only due to the GitHub file size limit). These two reports were generated in the same batch, but the pdf files were rendered differently. One has the code scrambled on the first page, the other on the 2nd page. It doesn't feel systemic.

[pdf_conversion_bug_maps_publish.zip](https://github.com/user-attachments/files/17499162/pdf_conversion_bug_maps_publish.zip)

**Comment by jdduh** [comment]
The same issue with the pdf reports after the Chinese language pack uninstalled.

**Comment by lbross** [comment]
Recommend that we leave PRISM buffer and units on this UI. It's here because the tool creates the prism aoi boundary files as part of the process. If we want to move that to the batch tool when we clip the prism boundaries, we could.
Do we need to support file geodatabase for the dem and the gauge station layers? If yes, do we need to support file geodatabase for any of the other layers?

**Comment by jdduh** [comment]
I suggest that we remove the PRISM, SNOTEL, and Snow Course clipping routines from the AOI creation process. We only need inputs for the source DEM (there is no need to differentiate 10 or 30 meters sources) and gauge stations (needed during BASIN creation, but not for Create AOI from a shapefile tool). We will rework the BAGIS Pro's Clip Layers interface so that users can clip more specific layers to the AOI if necessary.

**Comment by lbross** [comment]
How to generate a pourpoint: Zonal statistics (max statistic) as table run against flow accumulation with AOI boundary as zone. Returns cell with highest flow accumulation value. Choose one pixel. Convert to point becomes the pourpoint in the pourpoint file. Use Raster Calculator. Set max cell value to 1. Only one row allows in pourpoint vector.

**Comment by lbross** [comment]
What is the easiest way get local copies of the dem and gauge stations so that I can support dem and gauge stations layers in a File Geodatabase. I used to have a copy of the BAGIS DAta bin on an external drive, but that drive is no longer functional. This request could include other layer types in the BAGIS Data bin if you want to keep supporting file gdb with clipping

**Comment by jdduh** [comment]
The easiest way to get a DEM in a FGDB is using the surface.gdb in an existing AOI. We allow AOIs to be created within an AOI to capture the conceptual structure of nested watersheds. The gauge stations FGDB can be downloaded directly from https://nwcc-nrcs.hub.arcgis.com/ as a FGDB. I can create some test data if you have a preferred area (shapefile).
![image](https://github.com/user-attachments/assets/ccb9febb-02ed-4c9a-9a34-de4dc3b723fe)

**Comment by lbross** [comment]
If you could point me to an AOI with a subwatershed in our archives, that would be sufficient. I can use the aoi_v for the sub watershed with the dem from the parent watershed. This would also provide a source for the other layers from the parent if I need them eventually.

**Comment by jdduh** [comment]
The test data are available at D:\AOIs\2024_NWCC_AOI_Creation_Data on vsc-b1-11-1. The data include the Animas AOI and its subAOI (Animas_SubAOI_11062024). The shapefile of the subAOI is in the shapefiles folder. The gauge stations data is in the stations_FCST_ACTIVE_11062024 FGDB.
![image](https://github.com/user-attachments/assets/ef1b5b8f-9bf1-4bfd-8046-6b722d558105)

**Comment by lbross** [comment]
How important is it that the ArcGIS Pro output match the BAGIS v3 output? The numbers are close, but not exact. I think this may be part of the difference. The raster to polygon tool simplifies the polygons by default. BAGIS v3 didn't do this. The red line is pro and the black line is v3. This is a sample of the aoi_v file.
![image](https://github.com/user-attachments/assets/f9931203-112e-406b-9ef9-0b11b73575c7)

**Comment by jdduh** [comment]
I believe there is an option to not generalize the outline. Please enable it. It's critical to not generalize the AOI boundaries and that the "snap raster" option is activated when clipping the source DEM so the that there won't be any slivers when merging neighboring AOI polygons.

**Comment by lbross** [comment]
Yes. We do have the option NOT to simplify when converting the raster to polygon. I will make that change.
What snap raster should we use when clipping the source DEM? Throughout BAGIS-PRO we use the filled_dem as the snap raster but we don't have that yet when we are clipping the source DEM!

I have run into a problem with the flow accumulation tool. The tool runs and the range of values looks close to BAGIS v3, but when the raster is displayed, I don't see the channel like I do with the BAGIS v3 output. I posted a flow_accumulation.gdb.zip in the For_Geoffrey folder on basins that has the surfaces.gdb output up until this step.

**Comment by jdduh** [comment]
We use the source DEM in the settings file as the raster to be snapped to. This option is activated when a user selects a BASIN folder - given that almost all AOIs were created based on a BASIN. The flow accumulation raster looks correct. We need to change its symbolgy to the "histogram equalize" stretch type and set the Statistics to DRA (dynamic range adjustment) in BAGIS Pro whenever a flow_acc layer is added to the map.
![image](https://github.com/user-attachments/assets/f863519a-759e-446c-b631-17cf08742af0)

**Comment by jdduh** [comment]
This is the only AOI (out of almost 600) that has the buffer issue in both ArcMap (and probably ArcGIS Pro too). It's obvious that the buffer was created using just one of the dangle polygons. This has not been an issue for other AOIs that have multiple aoi_v polygons. Some of the reclipped layers generated by BAGIS Pro also have the wrong extent. These include the road and land cover layers (and maybe some others).

**Comment by lbross** [comment]
Upon further examination, it appears that the aoib_v is used extensively for clipping if the requested buffer distance is the same as the aoi buffer distance. It makes sense that I did this because it saves processing time not having to create the buffer file for every clip. How do you want to handle it? Is it okay to buffer aoib_v by 0.5m to get rid of the dangles and then re-save the file so we don't have to do this every time we clip to aoib_v?

Also I found that there is some old code that overwrites p_aoi_v with aoib_v if there is no prism buffer distance selected. We haven't run into it because we always buffer prism, but this should probably come out. I will regenerate p_aoi_v with the correct buffer distance independent of aoib_v. But we have the same question here, is we can modify the cartography slightly and persist it to avoid the dangles.

**Comment by lbross** [comment]
We agreed at our December 2 2024 call that we will take no action on aoib_v because it is only a problem with this one AOI.

**Comment by lbross** [comment]
I'm generating the pourpoint class but noticed that BAGIS v3 adds/populates several fields in this feature class that we don't use yet in bagis-pro. I created a [google doc](https://docs.google.com/spreadsheets/d/1xkfJ3sH2Ooq954TUtoXNxmyLTghet2wEzPelFSYaPII/edit?usp=sharing) with the fields for both this and aoi_v because some of the fields are shared. The ones in yellow are the ones that I am uncertain about.

**Comment by jdduh** [comment]
Additional remarks added to the highlighted fields in the shared Google sheet.

**Comment by lbross** [comment]
Thank you. 2 follow-up questions.

1. AOISHPAREA AND AOISHPUNIT. Your comment says 'We can switch to the new attribute fields if the UI can display the area'. What does this mean? Will we be creating a BAGIS-Pro version of AOI Utilities? We can store the area in SQKM (the AOISHPUNIT) and convert it to the other units for display if we need to.
2. awdb_id. Your comments says 'We can remove it for now, until we find its dependency in the rest of BAGIS Pro during the QA/QC stage.' Does this mean we want to eliminate this field from BAGIS-Pro? It shouldn't be too hard to do this because I use the field as a constant I can search for. I just haven't done it yet because it's been easier to leave it in.

**Comment by jdduh** [comment]
We need to create some types of AOI browser for the BAGIS Pro. See below for some screenshots from BAGIS V3.
If you are sure that BAGIS Pro won't use awdb_id (which is replaced by stationTriplet), then we can remove the awdb_id field.

![image](https://github.com/user-attachments/assets/879e3ccc-3b73-44fd-a5e5-9f079ed66b26)

![image](https://github.com/user-attachments/assets/1432bfcc-cb18-4053-ac23-23908f7061a8)

**Comment by lbross** [comment]
I have excised the awdb_id field from BAGIS-Pro. I opened two new issues for migrating the screens associated with the screenshots. The units for AOISHPAREA are stored in AOISHPUNIT so it should be easy to convert the area to other units when we need to. For now, I will only support SQKM since we control and populate this field, but the infrastructure is there if we need to add other units later.

**Comment by lbross** [comment]
For the hillshade layer, in addition to the z factor it also needs an azimuth and an altitude. The default values are 315 and 45 respectively. Is this okay?

**Comment by jdduh** [comment]
The default values for the hillshade tool are correct.

**Comment by jdduh** [comment]
The AOISHPAREA value needs to be updated every time BAGIS Pro retrieves its value. In theory, BAGIS Pro can access the shapearea value directly and check the spatial reference metadata (projection) to figure out the unit of the shapearea value. The reason that the original BAGIS maintains the AOISHPAREA and AOISHPUNIT is because the original Weasel AOI uses shapefile to store vector data. A shapefile doesn't have the shapearea field by default. After an AOI is converted to the fileGeodatabase format, all vector data are stored as featureclasses that by default have the shapearea field.

If AOISHPAREA is the source of the AOI area data in BAGIS Pro, then it needs to be updated every time BAGIS Pro retrieves its value. We current have several AOI reports (and summary statistics) that contain incorrect AOI area information due to that the AOISHPAREA values were not updated (due to the modification of the AOI boundary) or were missing (due to unknown errors when the AOI were created in BAGIS V3). It would be ideal to remove the AOISHPAREA dependency that needs to be checked by the user when they modify the AOI geometry. It's not a uncommon scenario to have the AOI geometry modified, particularly in large AOIs that encompass a diverse geographic regions.

**Comment by lbross** [comment]
Stop creating and populating aoishparea and aoishpunits field. No longer need to calculate AOI area.

**Comment by lbross** [comment]
Created a function called CalculateAoiAreaSqMetersAsync() that calculates the area of all polygons in aoi_v and returns that value in SqM. This function gets the Spatial Reference of the aoi_v feature class and queries the Unit which is a LinearUnit. Then we check the FactoryCode of the Unit to see if we are operating in Meters or Feet. A list of the factory codes is [here](https://pro.arcgis.com/en/pro-app/3.3/sdk/api-reference/topic8349.html). Currently we are only supporting 9001 and 9002. We can add others if needed.
Another option is to query the name of the Unit which appears to be 'Meter' or 'Foot' or ?. I can't find an enumeration of constants to restrict this to. That makes me uncomfortable because ESRI can change the value of the string and we will probably miss it. But it is an option.
aoishparea and aoishpunits fields have been purged from BAGIS-Pro. We no longer calculate/save the AOI area when creating an AOI with this tool.

**Comment by jdduh** [comment]
Here are some examples of the linear unit on ArcGIS Pro UI. Probably just check if the word "meters" appears in the Unit Name and ignore the conversion unit, i.e., "(1.0)".

![image](https://github.com/user-attachments/assets/61bcfbce-58f4-4db4-8b21-b54051d129f1)
![image](https://github.com/user-attachments/assets/a50e7810-2df1-422a-9f01-98cf877b1068)
![image](https://github.com/user-attachments/assets/34e919b7-095e-4b59-bfe1-a24afcc49d8e)

**Comment by lbross** [comment]
The properties of the SpatialReference.Unit are different than the display. It sounds like you want to use the SpatialReference.Unit.Name instead of the SpatialReference.Unit.FactoryCode which is fine. In the debugger I can see that two possible values are Meter and Foot. I will calculate the area based on these two constants. If ESRI changes the string values for name, hopefully we can catch it easily.
We talked about writing a log entry if the value isn't Meter. I haven't done this because it would be a little more work with the way I've currently got it implemented. Please let me know if you want me to add the logging.

**Comment by jdduh** [comment]
If BAGIS Pro can retrieve the area value correctly in the reports or summary tables, then we don't need to worry about logging it. It seems that the way you implemented the area unit conversion can take care of source vector layers with different linear units. If the  unit conversion is done by simple arithmetic calculation with the assumption that the source layer's linear unit is meter, then we will need to alert the users that the area value in reports/summary tables might be wrong. This feature is not urgent for now because all NWCC AOIs have the same spatial reference.

**Comment by lbross** [comment]
BAGIS-Pro can retrieve the area value correctly of aoi_v if the SpatialReference.Unit.Name is either 'Meter' or 'Foot'. If it is neither of these values it will return a -1. My concern is that the remainder of BAGIS-Pro expects aoi_v to be in Meters. We could make it support either meters or feet throughout, but this would be a lot of extra work when 95% of the time aoi_v will be in meters.
I understand the need to work with AOI's that may be in a different projection. I propose handling this during the AOI creation process. We can reproject aoi_v to Albers before proceeding with the rest of the AOI creation process (if it's not in meters).
I found a way to return multiple values from an async method so I can determine both the area and if aoi_v is in meters. From there it was pretty easy to add a log statement. I think we should log it on the off chance that an aoi_v isn't in meters.

**Comment by jdduh** [comment]
BAGIS V3 was used outside the US. In these cases, the Albers projection might not be the best spatial reference system. We will need a separate knowledge base to determine an appropriate projection system. BAGIS Pro's batch tool needs HUC2 values to function,

**Comment by jdduh** [comment]
BAGIS V3 allows users to customize its data sources. If we want to relax the geographic limitation of BAGIS Pro, then we need to have customizable data sources. We can keep this consideration for future BAGIS Pro improvement.

**Comment by lbross** [comment]
It looks like the BAGIS v3 version of this tool also clipped the sites layers and the prism layers if a GenerateAOIOnly flag was not set. I think we agreed to leave those tasks to the batch tool. Please confirm.

**Comment by jdduh** [comment]
That's correct. V3 has a function to automatically clip other layers to the AOI. We can remove that function from the AOI creation workflow, but we need to have a separate UI for users to add vector/raster to the AOI interactively. We can discuss this when we are about to modify the interactive tool's UI.

![Image](https://github.com/user-attachments/assets/19df9783-4073-49ff-b679-1a29e6b8341c)

**Comment by lbross** [comment]
I've been working on the progress indicator today for this tool. It's a little clunky but it's better than nothing. One thing that might help is to add the layers to the display at the end. I think the map window and the progress indicator may be competing for attention. Would this be okay if it works better? If it's not, I won't bother trying it out.

**Comment by lbross** [comment]
The progress indicator works better when we aren't adding the layers as they are created. I will adjust the UI and post an image when finished. The screen captures above are for the 'Generate AOI Tool' and should be under issue #58. This issue is for creating an AOI from a shapefile and the soon to be obsolete UI does allow users to select which layers to display.
